File: clients/java/src/main/java/cb/CustomWebsocketClient.java
package cb;

import cb.bots.BotAi;
import com.fasterxml.jackson.core.JsonParseException;
import com.fasterxml.jackson.databind.ObjectMapper;
import org.java_websocket.client.WebSocketClient;
import org.java_websocket.handshake.ServerHandshake;

import java.net.URI;
import java.nio.ByteBuffer;
import java.nio.charset.StandardCharsets;
import java.util.Map;

public class CustomWebsocketClient extends WebSocketClient {
    private final ObjectMapper mapper = new ObjectMapper();
    private final BotAi bot;

    public CustomWebsocketClient(URI serverUri, BotAi ai) {
        super(serverUri);
        bot = ai;
    }

    @Override
    public void onOpen(ServerHandshake handshakedata) {
        System.out.println("Connection to Four Connect Xtreme opened");
    }

    @Override
    public void onMessage(String s) {
        System.out.println("on message string");
    }

    @Override
    public void onMessage(ByteBuffer bytes) {
        System.out.println("on message bytes");
        try {
            String decodedMessage = new String(bytes.array(), StandardCharsets.UTF_8)
                    .replace("'", "\"");
            System.out.println(decodedMessage);
            PlayState response = mapper.readValue(decodedMessage, PlayState.class);
            if (response.getBot().equals(bot.getName())) {
                Map<String, Object> responseMap = Map.of(
                        "state", "play",
                        "column", bot.play(response)
                );
                String jsonResponse = mapper.writeValueAsString(responseMap);
                send(jsonResponse);
            }
        } catch (Exception e) {
            if (e instanceof JsonParseException) {
                System.out.println("Got ping message");
            } else {
                e.printStackTrace(System.out);
            }
        }
    }


    @Override
    public void onClose(int code, String reason, boolean remote) {
        System.out.println("Verbindung zum Server geschlossen: " + reason);
    }

    @Override
    public void onError(Exception ex) {
        System.err.println("Fehler: " + ex.getMessage());
    }
}

File: clients/dart/custom_websocket_client.dart
import 'dart:convert';
import 'dart:typed_data';
import 'package:web_socket_channel/web_socket_channel.dart';
import 'bots/bot_ai.dart';
import 'play_state.dart';

class CustomWebsocketClient {
  final Uri serverUri;
  final BotAi bot;
  late WebSocketChannel _channel;

  CustomWebsocketClient(this.serverUri, this.bot);

  void connect() {
    _channel = WebSocketChannel.connect(serverUri);
    print("Connection to Four Connect Xtreme opened");

    _channel.stream.listen(
          (message) {
        _onMessage(message);
      },
      onDone: () {
        print("Verbindung zum Server geschlossen");
      },
      onError: (error) {
        print("Fehler: $error");
      },
    );
  }

  void _onMessage(dynamic message) {
    if (message is! Uint8List) {
      print("on message string (ignoring): $message");
      return;
    }

    try {
      String decodedMessage = utf8.decode(message)
          .replaceAll("'", "\"");
      print(decodedMessage);

      PlayState response = PlayState.fromJson(jsonDecode(decodedMessage));

      if (response.bot == bot.getName()) {
        final responseMap = {
          "state": "play",
          "column": bot.play(response),
        };
        String jsonResponse = jsonEncode(responseMap);
        _channel.sink.add(jsonResponse);
      }
    } catch (e, s) {
      if (e is FormatException) {
        print("Got ping message");
      } else {
        print(e);
        print(s);
      }
    }
  }
}
